esp-bootloader-esp-idf/src/ota.rs

```diff
@@ -0,0 +1,463 @@
+        self.flash.read(SLOT1_DATA_OFFSET, buffer2.as_bytes_mut())?;
+        let seq0 = buffer1.ota_seq;
+        let seq1 = buffer2.ota_seq;
+        Ok((seq0, seq1))
+    }
+
+    /// Sets the currently active OTA-slot.
+    ///
+    /// Passing [Slot::None] will reset the OTA-data
+    pub fn set_current_slot(&mut self, slot: Slot) -> Result<(), crate::partitions::Error> {
+        if slot == Slot::None {
+            self.flash.write(SLOT0_DATA_OFFSET, &[0xffu8; 0x20])?;
+            self.flash.write(SLOT1_DATA_OFFSET, &[0xffu8; 0x20])?;
+            return Ok(());
+        }
+
+        let (seq0, seq1) = self.get_slot_seq()?;
+
+        let new_seq = {
+            if seq0 == 0xffffffff && seq1 == 0xffffffff {
+                1
+            } else if seq0 == 0xffffffff {
+                seq1 + 1
+            } else if seq1 == 0xffffffff {
+                seq0 + 1
+            } else {
+                u32::max(seq0, seq1) + 1
+            }
+        };
+
+        let crc = crate::crypto::Crc32::new();
+        let checksum = crc.crc(&new_seq.to_le_bytes());
+
+        let mut buffer = OtaSelectEntry::default();
+        self.flash.read(slot.offset(), buffer.as_bytes_mut())?;
+        buffer.ota_seq = new_seq;
+        buffer.crc = checksum;
+        self.flash.write(slot.offset(), buffer.as_bytes_mut())?;
+
+        Ok(())
+    }
+
+    /// Set the [OtaImageState] of the currently selected slot.
+    ///
+    /// # Errors
+    /// A [crate::partitions::Error::InvalidState] if the currently selected
+    /// slot is [Slot::None]
+    pub fn set_current_ota_state(
+        &mut self,
+        state: OtaImageState,
+    ) -> Result<(), crate::partitions::Error> {
+        match self.current_slot()? {
+            Slot::None => Err(crate::partitions::Error::InvalidState),
+            _ => {
+                let offset = self.current_slot()?.offset();
+                let mut buffer = OtaSelectEntry::default();
+                self.flash.read(offset, buffer.as_bytes_mut())?;
+                buffer.ota_state = state;
+                self.flash.write(offset, buffer.as_bytes_mut())?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Get the [OtaImageState] of the currently selected slot.
+    ///
+    /// # Errors
+    /// A [crate::partitions::Error::InvalidState] if the currently selected
+    /// slot is [Slot::None]
+    pub fn current_ota_state(&mut self) -> Result<OtaImageState, crate::partitions::Error> {
+        match self.current_slot()? {
+            Slot::None => Err(crate::partitions::Error::InvalidState),
+            _ => {
+                let offset = self.current_slot()?.offset();
+                let mut buffer = OtaSelectEntry::default();
+                self.flash.read(offset, buffer.as_bytes_mut())?;
+                Ok(buffer.ota_state)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::partitions::PartitionEntry;
+
+    struct MockFlash {
```

Comment:
I'd like to see a few tests (or all these tests) that use a `save-image` image from espflash that has a few partitions setup, instead of only the hard coded data we have here.

Reply:
Maybe we can do this later, but I'd like an issue for it if we're not doing it here.

Reply:
I don't really understand this one t.b.h.

We are only dealing with the OTA-data-partition here which is 2 x OtaSelectEntry spread to two sectors (i.e. 64 bytes of relevant data). I don't think espflash does anything relevant here (besides initializing OTA-data to 0xff)

Reply:
I guess I'm looking at the perspective of having to update these tests in the future, the random numbers are a bit scary, but if we had simple images that we tested on it might make things more approachable. I'm probably worrying about this a bit early though, so let's leave this for now.

Reply:
I see - I don't expect the hardcoded test data to become a problem, and I see how my comment wasn't that helpful

I don't think espflash can do anything regarding ota-data-partitions (other than erasing it).
We _could_ have a script using `otatool.py` to prepare testdata (and saving it to files with either espflash or esptool.py)

I'd say we should do that whenever we see a need for (much) more test data here